### ISSUE TITLE
Draft standard breakpoints and max widths

### ISSUE THREAD

**Issue by tylersticka** [cloudfour/cloudfour.com-patterns#547]
Our pattern library has traditionally included standard breakpoints, which are helpful for the reasons [I describe in this article](https://cloudfour.com/thinks/breakpoints-dont-box-me-in/). As of today, these have been:

```css
@custom-media --sm-viewport (width >= 30em);
@custom-media --md-viewport (width >= 48em);
@custom-media --lg-viewport (width >= 64em);
@custom-media --xl-viewport (width >= 80em);
@custom-media --xxl-viewport (width >= 100em);
@custom-media --xxxl-viewport (width >= 120em);
```

It has also included patterns for establishing maximum widths that prevent content from becoming too wide, either to maintain the integrity of the design or to maintain readability:

```css
.u-containSpread {
  max-width: 80em; /* ~ 1600 @ 20px base */
}

.u-containProse {
  max-width: 40rem; /* ~ 640 @ 16px base */
}
```

Historically, these have been thought of as separate decisions. But I suspect they should be thought of together: Not only should the prose width cleanly divide into the maximum spread/layout width, but both of those should have some relationship to our standard breakpoints.

## Tasks

- [ ] Define a set of standard breakpoints
- [ ] Define a set of maximum content widths for prose and layout
- [ ] Translate these into design token files
- [ ] Update existing CSS breakpoints (currently used for typography) to leverage token values

We can tweak these later, but we need a starting point!

**Comment by dromo77** [comment]
Going to tackle this next because I think it will be helpful for completing #481 .

**Comment by dromo77** [comment]
@tylersticka I noticed the current `.u-containSpread` utility not only establishes a maximum width, but also adds a horizontal margin at smaller screen sizes. Is that something the new utility should also do?

**Comment by tylersticka** [comment]
@dromo77 That's a great question.

Before I answer it, I want to clarify that the intent of this issue was to simply _define_ and _document_. I did not expect any new CSS patterns to emerge from it.

I don't want to _discourage_ over-achieving and introducing some containment patterns alongside that stuff, I just thought I'd mention it in case you read my answer and think to yourself "How is this a size 1 task??"

With that disclaimer out of the way, let's get back to your question:

> I noticed the current `.u-containSpread` utility not only establishes a maximum width, but also adds a horizontal margin at smaller screen sizes. Is that something the new utility should also do?

This exposes an issue I have with our current `u-contain*` utilities: I don't think they're really utilities! Utilities are meant to do _one thing_ and _only one thing_.

Sometimes, that _one thing_ involves multiple properties. Hiding things in an accessible way is complicated, for example. But in this case, I think the utility is actually doing _two_ things:

- ✅ Containing content to a maximum width
- ❌ Centering content horizontally

I marked that second part with a ❌ because there's nothing in the class name that would let you know that's what it's going to do. That's just something we _decided_ it should do.

And at the time, I'm sure we probably thought "there isn't a use case for _not_ doing that." But I've worked on a bunch of projects now where eventually you want something _next_ to the container, in which case you want the container to be aligned to that adjacent content, and you end up with silly things like this:

```html
<div class="u-containProse u-spaceRightNone">
  I'm contained
</div>

<div class="u-containProse">
  I'm contained and centered
</div>
```

So if this function is not a utility, the question becomes: What is it? And I think there are three main possibilities.

## 1. Two utilities

The simplest evolution of what we have already would be to use two utilities. This might end up looking like this:

```html
<div class="u-max-width-prose">
  I'm contained
</div>

<div class="u-max-width-prose u-space-horizontal-auto">
  I'm contained and centered
</div>
```

The nice thing about this approach is that the complexity of the `class` attribute is proportionate to the complexity of the styles we're applying. We aren't adding classes to _override_ others, and we aren't applying styles we wouldn't expect a utility to apply.

The bad thing about this approach is that it's kind of clunky to re-use, and it makes it hard to apply opinionated design touches like whitespace, etc.

## 2. An object pattern

[In his article on namespaces](https://csswizardry.com/2015/03/more-transparent-ui-code-with-namespaces/), Harry Roberts describes the "objects" namespace as the best place for "things like layout, wrappers and containers." They are more opinionated than utilities, but they are nothing without content to fill them.

In that case, we can go ahead and apply opinionated styles because that's what it's meant for.

```html
<div class="o-container o-container--prose">
  I'm contained and centered by default
</div>

<div class="o-container o-container--prose o-container--align-left">
  I'm contained and left-aligned
</div>
```

In this case, we could optionally add modifiers to expose those different variations we need in terms of how they impact the object pattern (rather than disconnected, one-off things).

### 3. A combination of both

These techniques are not mutually exclusive. We could create an opinionated `o-container` object that centers by default, and if we only need it to _not_ be centered once or twice, we can override it:

```html
<div class="o-container o-container--prose">
  I'm contained and centered by default
</div>

<div class="o-container o-container--prose u-space-right-initial">
  I'm contained and left-aligned
</div>
```

## TL;DR

- You don't have to make a container pattern for this task if you don't want to.
- If you do, I would probably make it an _object_ pattern.
- Making it an object pattern frees us up to make opinionated choices now, alter them later, and add utilities or modifiers as we see fit.

Reach out if you want to talk through any of what I just described. 🙂

**Comment by dromo77** [comment]
Thanks @tylersticka! I think I will create a separate issue for a container pattern instead of including it here. That being the case, what do you think is the best way to document the maximum widths for prose and layout in the meantime?

**Comment by tylersticka** [comment]
@dromo77 Hmm, I'm having trouble clarifying differently than I did in the "Tasks" section of this issue. Is there a particular one that seems open-ended or unclear?

**Comment by dromo77** [comment]
> Define a set of maximum content widths for prose and layout

@tylersticka I originally read this and thought it meant this issue encompassed our `u-contain*` utilities. After rereading the tasks, is the goal to create tokens for prose and layout max-widths?

**Comment by tylersticka** [comment]
@dromo77 The goal is to create design tokens for breakpoints _and_ max-widths, since the two inevitably co-exist in some fashion.

An example of what _not_ to do from our current patterns: We have breakpoints for `30em` and `48em`, but our prose max-width is `40em`.

**Comment by dromo77** [comment]
@tylersticka Thanks for the clarification, the part I was missing was creating tokens for max-widths in addition to breakpoints.
